### PR TITLE
Reduce: FPGA expansion

### PR DESCRIPTION
**Pull request by TizianoDeMatteis**
This PR adds an FPGA-specific expansion for Reduce.

### CODE REVIEW

File: dace/libraries/standard/nodes/reduce.py [hunk 1]
```diff
@@ -804,6 +803,243 @@ def expansion(node: 'Reduce', state: SDFGState, sdfg: SDFG):
         #return reduce_node.expand(sdfg, state)
 
 
+@dace.library.expansion
+class ExpandReduceFPGA(pm.ExpandTransformation):
+    """
+        FPGA SDFG Reduce expansion. This does not assume special hardware future (such as sigle clock-cycle
+        accumulation).
```
Comment (definelicht) [inline dace/libraries/standard/nodes/reduce.py:810]:
```suggestion
        FPGA SDFG Reduce expansion. This does not assume single-cycle accumulation of the given data type.
```

File: dace/libraries/standard/nodes/reduce.py [hunk 2]
```diff
@@ -804,6 +803,243 @@ def expansion(node: 'Reduce', state: SDFGState, sdfg: SDFG):
...
+        # Accumulate over partial results
+        redtype = detect_reduction_type(node.wcr)
+        if redtype not in ExpandReduceFPGA._REDUCTION_TYPE_EXPR:
+            raise ValueError('Reduction type not supported for "%s"' % node.wcr)
+        else:
+            reduction_expr = ExpandReduceFPGA._REDUCTION_TYPE_EXPR[redtype]
+
+        # generate flatten index considering inner map: will be used for indexing into partial results
+        ranges_size = ime.range.size()
+        inner_index = '{}'.format('+'.join([
```
Comment (definelicht) [inline dace/libraries/standard/nodes/reduce.py:949]: Mix of different string format syntax... Why not stick to f-strings? 🤓

File: dace/libraries/standard/nodes/reduce.py [hunk 3]
```diff
@@ -804,6 +803,243 @@ def expansion(node: 'Reduce', state: SDFGState, sdfg: SDFG):
...
+            '_i%d * %d' % (i, ranges_size[i + 1]) for i in range(len(axes) - 1)
+        ]))
+        inner_index = inner_index + "{}_i{}".format(
```
Comment (definelicht) [inline dace/libraries/standard/nodes/reduce.py:952]: Even a third type of string formatting here!

File: dace/libraries/standard/nodes/reduce.py [hunk 4]
```diff
@@ -804,6 +803,243 @@ def expansion(node: 'Reduce', state: SDFGState, sdfg: SDFG):
...
+                               dst_conn="reduce_in",
+                               memlet=dace.Memlet(f"{reduce_name}[0]"))
+        nstate.add_memlet_path(reduce_tasklet,
+                               reduce_exit,
+                               reduce_access,
+                               src_conn="reduce_out",
+                               memlet=dace.Memlet(f"{reduce_name}[0]"))
+
+        if not all_axis:
+            #write out the rsult
```
Comment (definelicht) [inline dace/libraries/standard/nodes/reduce.py:1028]:
```suggestion
            # Write out the result

```

File: tests/fpga/reduce_fpga.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+                   storage=dace.dtypes.StorageType.FPGA_Global,
+                   transient=True)
+
+    in_device_A = copy_in_state.add_write("device_A")
+
+    copy_in_memlet = dace.Memlet("A[{}]".format(",".join(
+        ['0:%d' % i for i in input_data_shape])))
+
+    copy_in_state.add_memlet_path(in_host_A, in_device_A, memlet=copy_in_memlet)
+    ###########################################################################
```
Comment (definelicht) [inline tests/fpga/reduce_fpga.py:45]:
```suggestion
    copy_in_state.add_memlet_path(in_host_A, in_device_A, memlet=copy_in_memlet)
    
    ###########################################################################
```

File: tests/fpga/reduce_fpga.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    sdfg.add_array("device_B",
+                   shape=output_data_shape,
+                   dtype=dtype,
+                   storage=dace.dtypes.StorageType.FPGA_Global,
+                   transient=True)
+
+    out_device = copy_out_state.add_read("device_B")
+    out_host = copy_out_state.add_write("B")
+    copy_out_memlet = dace.Memlet("B[{}]".format(",".join(
+        ['0:%d' % i for i in output_data_shape])))
```
Comment (definelicht) [inline tests/fpga/reduce_fpga.py:59]: Mix of two kinds of format strings... In one expression!! :D

File: tests/fpga/reduce_fpga.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    sdfg.validate()
+
+    return sdfg
+
+
+def test_reduce_sum_one_axis():
+    A = np.random.rand(8, 8).astype(np.float32)
+    B = np.random.rand(8).astype(np.float32)
+    sdfg = create_reduce_sdfg('lambda a,b: a+b', [0], "reduction_sum_one_axis",
+                              A, B)
```
Comment (definelicht) [inline tests/fpga/reduce_fpga.py:90]: It also hurts my eyes a bit that single quotes and double quotes are used completely interchangeably ;-) Stick to one?

File: tests/xilinx_test.sh [hunk 8]
```diff
@@ -119,6 +119,9 @@ run_all() {
     run_sample blas/nodes/gemv_test gemv_FPGA_Accumulate_float_False_w4_1 1 --target accumulate --vectorize 4
     run_sample blas/nodes/ger_test ger_test_1 1 --target fpga
 
+    ## STD LibNodes
+    run_multi_sample fpga/reduce_fpga 0 "reduction_sum_one_axis" "reduction_sum_all_axis" "reduction_sum_4D" "reduction_max"
```
Comment (definelicht) [inline tests/xilinx_test.sh:123]: In this case it's quite important to test that this has been pipelined successfully -- can you make sure it tests, e.g., sum with doubles, and checks that the partial product implementation pipelines this?

File: dace/libraries/standard/nodes/reduce.py [hunk 9]
```diff
@@ -804,6 +803,243 @@ def expansion(node: 'Reduce', state: SDFGState, sdfg: SDFG):
         #return reduce_node.expand(sdfg, state)
 
 
+@dace.library.expansion
+class ExpandReduceFPGA(pm.ExpandTransformation):
```
Comment (definelicht) [inline dace/libraries/standard/nodes/reduce.py:807]: What do you think about calling this something along the lines of `ExpandReduceFPGAPartialReduction`, so we can also have a more efficient non-partial version (in the future, if we want to)?

**Review by definelicht — CHANGES_REQUESTED** [review]
Nice work, just some small nitpicks :-)

**Review by definelicht — APPROVED** [review]
Very nice!

**Comment by TizianoDeMatteis** [comment]
Thanks and thanks for the comments!

**Pull request merged**